File: dndacademyBE/src/main/java/com/giulia/dndacademy/model/Character.java
package com.giulia.dndacademy.model;

import com.giulia.dndacademy.model.enumerations.AttackAbility;
import jakarta.persistence.*;
import lombok.*;

@Entity
@Table(name = "characters")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Character {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String name;

    private String race;

    private String characterClass;

    @Column(nullable = false)
    private int level;

    private int maxHp;

    private int currentHp;

    private int armorClass;

    private String weaponName;

    private int damageDie;

    @Enumerated(EnumType.STRING)
    private AttackAbility attackAbility;

    @Builder.Default
    @Column(nullable = false)
    private boolean spellcaster = false;

    private String spellName;

    private int spellDamageDie;

    @Enumerated(EnumType.STRING)
    private AttackAbility spellAbility;

    @Builder.Default
    @Column(nullable = false)
    private boolean alive = true;

    @ManyToOne
    @JoinColumn(name = "player_id")
    private User player;

    @ManyToOne
    @JoinColumn(name = "campaign_id")
    private Campaign campaign;

    @OneToOne(mappedBy = "character", cascade = CascadeType.ALL)
    private CharacterStats stats;
}

File: dndacademyfe/src/types/character.ts
export type AttackAbility =
  | "STRENGTH"
  | "DEXTERITY"
  | "INTELLIGENCE"
  | "WISDOM"
  | "CHARISMA"

export interface CharacterStats {
  strength: number
  dexterity: number
  constitution: number
  intelligence: number
  wisdom: number
  charisma: number
}

export interface CreateCharacterRequest {
  name: string
  race: string
  characterClass: string
  level: number
  campaignId: number
  maxHp: number
  armorClass: number
  weaponName: string
  damageDie: number
  attackAbility: AttackAbility
  spellcaster: boolean
  spellName: string | null
  spellDamageDie: number
  spellAbility: AttackAbility | null
  stats: CharacterStats
}

export interface UpdateCharacterRequest {
  name: string
  race: string
  characterClass: string
  level: number
  maxHp: number
  armorClass: number
  weaponName: string
  damageDie: number
  attackAbility: AttackAbility
  spellcaster: boolean
  spellName: string | null
  spellDamageDie: number
  spellAbility: AttackAbility | null
  stats: CharacterStats
}

export interface Character {
  id: number
  name: string
  race: string
  characterClass: string
  level: number
  maxHp: number
  currentHp: number
  armorClass: number
  playerUsername: string | null
  campaignId: number

  imageUrl: string | null

  weaponName: string
  damageDie: number
  attackAbility: AttackAbility

  spellcaster: boolean
  spellName: string | null
  spellDamageDie: number
  spellAbility: AttackAbility | null
  stats?: CharacterStats
}
